### ISSUE TITLE
Update client optimizers (moving clipping part before weights update)

### ISSUE THREAD

**Issue by Xinran-Zhao** [APPFL/APPFL#197]
**Description:**
I am looking through the client optimizers. Currently, the clipping part `torch.nn.utils.clip_grad_norm_` runs after the weights update `optimizer.step()`. I think the clipped gradient should run before the weights update, in order to be used in optimization. 
Thank you for addressing this concern.

**Code:**
https://github.com/APPFL/APPFL/blob/c1270c293b406f4efaae1211cc9d569243f3bd43/src/appfl/algorithm/legacy/client_optimizer.py#L41-L51

**Comment by Zilinghan** [comment]
Hi @Xinran-Zhao, thanks for bringing up this issue, and it is fixed by moving `optimizer.step()` after gradient clippong.
